src/json-observe.ts

```diff
@@ -0,0 +1,262 @@
+        }
+      });
+    return proxy;
+  }
+
+  //grab tree's leaves one by one, encapsulate them into a proxy and return
+  private _proxifyObjectTreeRecursively(root: any, path): any {
+    if(!path) path = "";
+    for (var key in root) {
+      if (root.hasOwnProperty(key)) {
+        var distPath = path + '/' + this.escapePathComponent(key);
+        if (typeof root[key] === 'object') {
+          root[key] = this.generateProxyAtPath(root[key], distPath);
+          this._proxifyObjectTreeRecursively(root[key], distPath);
+        }
+      }
+    }
+    return this.generateProxyAtPath(root, "/");
+  }
+
+  //this function is for aesthetic purposes
+  private proxifyObjectTree(root: any): any {
+    
+    /*
+    while proxyifying object tree,
+    the proxyifying operation itself is being
+    recorded, which in an unwanted behavior,
+    that's why we disable recording through this
+    inital process;
+    */
+
+    this.disableCallback = true;
+    var proxifiedObject = this._proxifyObjectTreeRecursively(root, "");
+
+    /* OK you can record now */
+    this.disableCallback = false;
+
+    return proxifiedObject;
+  }
+  private equalsEvents(op1, op2): boolean {
+    if (op1.value && op2.value)
+      return op1.op === op2.code && op1.path === op2.path && op1.value === op2.value;
+
+    return op1.op === op2.code && op1.path === op2.path;
+  }
+  constructor(root) {
+    this.originalObject = root;
+    this.cachedProxy = null;
+    this.patches = [];
+    this.isRecording = false;
+    this.userCallback;
+
+    this.defaultCallback = function (sender) {
+      return function (event) {
+        //don't duplicate exact same events
+        //if(sender.patches.length > 0 && this.equalsEvents(sender.patches[sender.patches.length - 1], event)) return;
+
+        if (!sender.disableCallback) {
+          if (sender.isRecording) {
+            sender.patches.push(event);
+          }
+          if (sender.userCallback) {
+            sender.userCallback(event);
+          }
+        }
+      }
+    } (this);
+  }
+  public observe(record, cb) {
+    if (!record && !cb) {
+      throw new Error('You need to either record changes or pass a defaultCallback');
+    }
+    this.isRecording = record;
+    if(cb) this.userCallback = cb;
+    return this.cachedProxy = this.proxifyObjectTree(deepClone(this.originalObject));
+  }
+  public generate() {
+    if (!this.isRecording) {
+      throw new Error('You should set record to true to get patches later');
+    }
+    /*
+    TODO: here, we could remove duplicates.
+    But we can't just UNINQE the array.
+    because consider [change1, change2, change1]
+    assuming all changes are on the same path,
+    both change1's are neccessary. 
+    */
+
+    return this.patches.splice(0, this.patches.length);
+  }
+  public unobserve(keepHistory) {
+    if (!keepHistory) this.patches = [];
+
+    //return a normal, non-proxified object
+    return deepClone(this.cachedProxy);
```

Comment:
After a second thought, yes, I can remove it, but what's the point of the whole function this way? It will be useless. Once the `observedObject` is returned to the user, we can't actually `unobserve` it from inside the class anymore. It's out of hand, so we offer `unobserve` function, begging the user to set `observedObject = unobserve()` this way, the proxified version can die and be garbage collected. 

Right?